packages/go/trynoports/app_update.go

```diff
@@ -0,0 +1,162 @@
+)
+
+func (m appState) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	var (
+		cmd  tea.Cmd
+		cmds []tea.Cmd
+	)
+
+	// Handle terminal / input events first
+	switch msg := msg.(type) {
+	case ProgramMsg:
+		m.program = msg.program
+	case CmdDoneMsg:
+		m.viewport.isRunning = false
+	case ViewportContentMsg:
+		m.viewport.content += msg.content
+		m.viewport.model.SetContent(m.viewport.content)
+		m.viewport.model.GotoBottom()
+		if useHighPerformanceRenderer {
+			viewport.Sync(m.viewport.model)
+		}
+	case tea.WindowSizeMsg:
+		m.width = msg.Width
+		m.height = msg.Height
+	case tea.KeyMsg:
+		m = m.KeyMsg(msg)
+	}
+
+	// Update the view contents
+	if m.viewport.isReady {
+		// Update viewport
+		m.viewport.model, cmd = m.viewport.model.Update(msg)
+		cmds = append(cmds, cmd)
+	} else {
+		m.viewport.spinner, cmd = m.viewport.spinner.Update(msg)
+		cmds = append(cmds, cmd)
+	}
+
+	// Update the list
+	m.list.model, cmd = m.list.model.Update(msg)
+	cmds = append(cmds, cmd)
+
+	// Recompute the size of the components based on current changes
+	m, cmd = m.ResizeComponents(msg)
+	cmds = append(cmds, cmd)
+
+	// Batch all of the potential commands
+	return m, tea.Batch(cmds...)
+}
+
+// Handle keyboard input
+func (p appState) KeyMsg(msg tea.KeyMsg) (m appState) {
+	m = p
+	switch msg.String() {
+	case "enter":
+		if m.viewport.isRunning || m.list.model.Index() == m.viewport.contentIndex {
+			return
+		}
+		m.viewport.isRunning = true
+		m.viewport.contentIndex = m.list.model.Index()
+
+		item := m.list.model.SelectedItem()
+		var command, args string
+
+		switch item.(type) {
+		case appCommand:
+			// Update the command entry
+			command = item.(appCommand).command
+			args = strings.Join(m.list.model.SelectedItem().(appCommand).args, " ")
+			// Throw away the old contents, otherwise someone malicious might try to allocate infinite memory to that struct
+			m.viewport.content = fmt.Sprintf("> %s %s\n", command, args)
+
+		default:
+			m.viewport.content = ""
+		}
+		m.viewport.model.SetContent(m.viewport.content)
+		m.viewport.model.GotoBottom()
+		if useHighPerformanceRenderer {
+			viewport.Sync(m.viewport.model)
+		}
+
+		done := make(chan int, 1)
+		ch, err := m.list.model.SelectedItem().(RunnerWithDone).Run(done)
+		if err != nil {
+			close(done)
+			// Welcome command should never return an error, so it is safe to assume we have an appCommand in here
+			// Thus command & args are set
+			m.viewport.content = fmt.Sprintf("> %s %s\nOops! We messed up, can't run this command right now...", command, args)
+			m.viewport.model.SetContent(m.viewport.content)
+			m.viewport.model.GotoBottom()
+			m.viewport.isRunning = false
+			log.Error("Error running command: ", "error", err)
+			return
+		}
+
+		go func() {
+			for {
+				select {
+				case <-done:
+					close(ch)
+					close(done)
+					m.program.Send(CmdDoneMsg{})
+					return
+				case msg := <-ch:
+					m.program.Send(ViewportContentMsg{msg})
+				}
+			}
+		}()
```

Comment:
asynchronously receive messages from the channel we opened, then send them back to bubbletea to display to the viewport, send CmdDoneMsg when the program is done